esp-hal/src/soc/mod.rs

```diff
@@ -83,36 +97,40 @@ impl self::efuse::Efuse {
 }
 
 #[allow(unused)]
-pub(crate) fn is_valid_ram_address(address: u32) -> bool {
-    (self::constants::SOC_DRAM_LOW..=self::constants::SOC_DRAM_HIGH).contains(&address)
+pub(crate) fn is_valid_ram_address(address: usize) -> bool {
+    addr_in_range(address, DRAM)
 }
 
 #[allow(unused)]
 pub(crate) fn is_slice_in_dram<T>(slice: &[T]) -> bool {
-    let start = slice.as_ptr() as u32;
-    let end = start + slice.len() as u32;
-    self::constants::SOC_DRAM_LOW <= start && end <= self::constants::SOC_DRAM_HIGH
+    slice_in_range(slice, DRAM)
 }
 
 #[allow(unused)]
-pub(crate) fn is_valid_psram_address(address: u32) -> bool {
-    #[cfg(psram)]
-    {
-        let memory_range = MAPPED_PSRAM.with(|mapped_psram| mapped_psram.memory_range.clone());
-        memory_range.contains(&(address as usize))
-    }
-    #[cfg(not(psram))]
-    false
+pub(crate) fn is_valid_psram_address(address: usize) -> bool {
+    addr_in_range(address, psram_range())
 }
 
 #[allow(unused)]
 pub(crate) fn is_slice_in_psram<T>(slice: &[T]) -> bool {
-    let start = slice.as_ptr() as u32;
-    let end = start + slice.len() as u32;
-    is_valid_psram_address(start) && is_valid_psram_address(end)
+    slice_in_range(slice, psram_range())
 }
 
 #[allow(unused)]
-pub(crate) fn is_valid_memory_address(address: u32) -> bool {
+pub(crate) fn is_valid_memory_address(address: usize) -> bool {
     is_valid_ram_address(address) || is_valid_psram_address(address)
 }
+
+fn slice_in_range<T>(slice: &[T], range: Range<usize>) -> bool {
+    let slice = slice.as_ptr_range();
+    let start = slice.start as usize;
+    let end = slice.end as usize;
+    // `end` is >= `start`, so we don't need to check that `end > range.start`
+    // `end` also points one after the last element, so we need to check that it's
+    // strictly less than `range.end`
```

Comment:
Is `range.end` also one after the last valid memory location? I'm failing to convince myself that `<` is more correct than `<=`

Reply:
I'm thinking this needs to be something to the effect of `end <= SOC_DRAM_HIGH + 1`.

Reply:
Yeah I looked through the DRAM constants and they are all exclusive, and PSRAM range is set up as start+length which is also exclusive.

That being said, I'm pretty sure I got this wrong (if I declare a slice that takes up all of DRAM, it needs to be contained within the DRAM range, so `end <= range.end` should probably be the correct one here. I'll revisit tomorrow.

Reply:
Thanks for noticing this.